pkg/controller/nodes/array/handler.go

```diff
@@ -0,0 +1,593 @@
+package array
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+
+	idlcore "github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/core"
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/event"
+
+	"github.com/flyteorg/flyteplugins/go/tasks/pluginmachinery/core"
+	"github.com/flyteorg/flyteplugins/go/tasks/pluginmachinery/ioutils"
+	"github.com/flyteorg/flyteplugins/go/tasks/plugins/array/errorcollector"
+
+	"github.com/flyteorg/flytepropeller/events"
+	"github.com/flyteorg/flytepropeller/pkg/apis/flyteworkflow/v1alpha1"
+	"github.com/flyteorg/flytepropeller/pkg/compiler/validators"
+	"github.com/flyteorg/flytepropeller/pkg/controller/config"
+	"github.com/flyteorg/flytepropeller/pkg/controller/executors"
+	"github.com/flyteorg/flytepropeller/pkg/controller/nodes/errors"
+	"github.com/flyteorg/flytepropeller/pkg/controller/nodes/handler"
+	"github.com/flyteorg/flytepropeller/pkg/controller/nodes/interfaces"
+	"github.com/flyteorg/flytepropeller/pkg/controller/nodes/task"
+	"github.com/flyteorg/flytepropeller/pkg/controller/nodes/task/k8s"
+
+	"github.com/flyteorg/flytestdlib/bitarray"
+	"github.com/flyteorg/flytestdlib/logger"
+	"github.com/flyteorg/flytestdlib/promutils"
+	"github.com/flyteorg/flytestdlib/storage"
+)
+
+var (
+	nilLiteral = &idlcore.Literal{
+		Value: &idlcore.Literal_Scalar{
+			Scalar: &idlcore.Scalar{
+				Value: &idlcore.Scalar_NoneType{
+					NoneType: &idlcore.Void{},
+				},
+			},
+		},
+	}
+)
+
+//go:generate mockery -all -case=underscore
+
+// arrayNodeHandler is a handle implementation for processing array nodes
+type arrayNodeHandler struct {
+	eventConfig                *config.EventConfig
+	metrics                    metrics
+	nodeExecutor               interfaces.Node
+	pluginStateBytesNotStarted []byte
+	pluginStateBytesStarted    []byte
+}
+
+// metrics encapsulates the prometheus metrics for this handler
+type metrics struct {
+	scope promutils.Scope
+}
+
+// newMetrics initializes a new metrics struct
+func newMetrics(scope promutils.Scope) metrics {
+	return metrics{
+		scope: scope,
+	}
+}
+
+// Abort stops the array node defined in the NodeExecutionContext
+func (a *arrayNodeHandler) Abort(ctx context.Context, nCtx interfaces.NodeExecutionContext, reason string) error {
+	arrayNode := nCtx.Node().GetArrayNode()
+	arrayNodeState := nCtx.NodeStateReader().GetArrayNodeState()
+
+	externalResources := make([]*event.ExternalResourceInfo, 0, len(arrayNodeState.SubNodePhases.GetItems()))
+	messageCollector := errorcollector.NewErrorMessageCollector()
+	switch arrayNodeState.Phase {
+	case v1alpha1.ArrayNodePhaseExecuting, v1alpha1.ArrayNodePhaseFailing:
+		currentParallelism := uint32(0)
+		for i, nodePhaseUint64 := range arrayNodeState.SubNodePhases.GetItems() {
+			nodePhase := v1alpha1.NodePhase(nodePhaseUint64)
+
+			// do not process nodes that have not started or are in a terminal state
+			if nodePhase == v1alpha1.NodePhaseNotYetStarted || isTerminalNodePhase(nodePhase) {
```

Comment:
should we also continue in case the node is already in a terminal state?

Reply:
This should be covered by the `isTerminalNodePhase` check here.